engine/servermaster/executor_manager.go

```diff
@@ -148,51 +163,56 @@ func (e *ExecutorManagerImpl) HandleHeartbeat(req *pb.HeartbeatRequest) (*pb.Hea
 	return resp, nil
 }
 
-// RegisterExec registers executor to both executor manager and resource manager
-func (e *ExecutorManagerImpl) RegisterExec(info *model.NodeInfo) {
-	log.Info("register executor", zap.Any("info", info))
+// registerExec registers executor to both executor manager and resource manager.
+// Note that this method must be called with the lock held.
+func (e *ExecutorManagerImpl) registerExecLocked(ormExecutor *ormModel.Executor) {
+	log.Info("register executor", zap.Any("executor", ormExecutor))
 	exec := &Executor{
-		NodeInfo:       *info,
+		Executor:       *ormExecutor,
 		lastUpdateTime: time.Now(),
 		heartbeatTTL:   e.initHeartbeatTTL,
 		status:         model.Initing,
 		logRL:          rate.NewLimiter(rate.Every(time.Second*5), 1 /*burst*/),
 	}
-	e.mu.Lock()
-	e.executors[info.ID] = exec
+	e.executors[ormExecutor.ID] = exec
 	e.notifier.Notify(model.ExecutorStatusChange{
-		ID:   info.ID,
+		ID:   ormExecutor.ID,
 		Tp:   model.EventExecutorOnline,
-		Addr: info.Addr,
+		Addr: ormExecutor.Address,
 	})
-	e.mu.Unlock()
-	e.rescMgr.Register(exec.ID, exec.Addr, model.RescUnit(exec.Capability))
+	e.rescMgr.Register(exec.ID, exec.Address, model.RescUnit(exec.Capability))
 }
 
 // AllocateNewExec allocates new executor info to a give RegisterExecutorRequest
 // and then registers the executor.
-func (e *ExecutorManagerImpl) AllocateNewExec(req *pb.RegisterExecutorRequest) (*model.NodeInfo, error) {
-	executor := req.Executor
-	log.Info("allocate new executor", zap.Stringer("executor", executor))
+func (e *ExecutorManagerImpl) AllocateNewExec(ctx context.Context, req *pb.RegisterExecutorRequest) (*ormModel.Executor, error) {
+	pbExecutor := req.Executor
+	log.Info("allocate new executor", zap.Stringer("executor", pbExecutor))
 
 	e.mu.Lock()
+	defer e.mu.Unlock()
```

Comment:
Do you think it would be a good idea to use a cancellable lock here?

Reply:
After a deep investigation, I found a cancellable lock would bring a lot of changes.

To reduce lock wait time, I move database operation out of lock held. If the database operation fails, the client shouldn't reuse the executor in memory map, which will be removed after ttl expires.